tests/test_cli/test_migrate_v3.py

```diff
@@ -0,0 +1,632 @@
+import lzma
+from pathlib import Path
+from typing import Literal, cast
+
+import numcodecs
+import numcodecs.abc
+import numpy as np
+import pytest
+from numcodecs.zarr3 import LZMA, Delta
+
+import zarr
+from tests.test_cli.conftest import create_nested_zarr
+from zarr.abc.codec import Codec
+from zarr.codecs.blosc import BloscCodec
+from zarr.codecs.bytes import BytesCodec
+from zarr.codecs.gzip import GzipCodec
+from zarr.codecs.transpose import TransposeCodec
+from zarr.codecs.zstd import ZstdCodec
+from zarr.core.array import Array
+from zarr.core.chunk_grids import RegularChunkGrid
+from zarr.core.chunk_key_encodings import V2ChunkKeyEncoding
+from zarr.core.common import JSON, ZarrFormat
+from zarr.core.dtype.npy.int import UInt8, UInt16
+from zarr.core.group import Group
+from zarr.storage._local import LocalStore
+
+typer_testing = pytest.importorskip(
+    "typer.testing", reason="optional cli dependencies aren't installed"
+)
+cli = pytest.importorskip("zarr._cli.cli", reason="optional cli dependencies aren't installed")
+
+runner = typer_testing.CliRunner()
+
+
+def test_migrate_array(local_store: LocalStore) -> None:
+    shape = (10, 10)
+    chunks = (10, 10)
+    dtype = "uint16"
+    compressors = numcodecs.Blosc(cname="zstd", clevel=3, shuffle=1)
+    fill_value = 2
+    attributes = cast(dict[str, JSON], {"baz": 42, "qux": [1, 4, 7, 12]})
+
+    zarr.create_array(
+        store=local_store,
+        shape=shape,
+        chunks=chunks,
+        dtype=dtype,
+        compressors=compressors,
+        zarr_format=2,
+        fill_value=fill_value,
+        attributes=attributes,
+    )
+
+    result = runner.invoke(cli.app, ["migrate", "v3", str(local_store.root)])
+    assert result.exit_code == 0
+    assert (local_store.root / "zarr.json").exists()
+
+    zarr_array = zarr.open(local_store.root, zarr_format=3)
+    metadata = zarr_array.metadata
+    assert metadata.zarr_format == 3
+    assert metadata.node_type == "array"
+    assert metadata.shape == shape
+    assert metadata.chunk_grid == RegularChunkGrid(chunk_shape=chunks)
+    assert metadata.chunk_key_encoding == V2ChunkKeyEncoding(separator=".")
+    assert metadata.data_type == UInt16(endianness="little")
+    assert metadata.codecs == (
+        BytesCodec(endian="little"),
+        BloscCodec(typesize=2, cname="zstd", clevel=3, shuffle="shuffle", blocksize=0),
+    )
+    assert metadata.fill_value == fill_value
+    assert metadata.attributes == attributes
+    assert metadata.dimension_names is None
+    assert metadata.storage_transformers == ()
```

Comment:
Instead of asserting every attribute on metadata, it might be nicer to do `assert metadata == MetadataObject(...)`. That way if multiple attributes are wrong, they will all show up in the pytest error, as opposed to with multiple asserts pytest will just stop on the first failure.